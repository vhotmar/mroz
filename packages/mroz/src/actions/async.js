import { createAction } from "redux-actions";

export const request = createAction("@@mroz/ASYNC_REQUEST", id => ({
  id
}));

export const success = createAction("@@mroz/ASYNC_SUCCESS", (id, data) => ({
  id,
  data
}));

export const fail = createAction("@@mroz/ASYNC_FAIL", (id, error) => ({
  id,
  error
}));
