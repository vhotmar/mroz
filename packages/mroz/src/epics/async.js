import R from "ramda";
import isFunction from "lodash/isFunction";
import { Observable } from "rxjs";
import asyncArc from "../arcs/async";

const emptyObservableFn = () => Observable.empty();
const identity = x => x;

const mergeObservableCreator = (a = emptyObservableFn) => (
  b = emptyObservableFn
) => (...args) => Observable.merge(b(...args), a(...args));

const makeObservableEdit = propName => newProp => (config = {}) =>
  R.over(R.lensProp(propName), mergeObservableCreator(newProp), config);

const makeComposeEdit = propName => (newProp = identity) => (config = {}) =>
  R.over(
    R.lensProp(propName),
    prop => (s, ...args) =>
      newProp(isFunction(prop) ? prop(s, ...args) : s, ...args),
    config
  );

export const addRequest = makeObservableEdit("request");

export const addSuccess = makeObservableEdit("success");

export const addFail = makeObservableEdit("fail");

export const addWrapper = makeComposeEdit("wrapper");

export const addFilter = makeComposeEdit("filter");

export const addNormalize = makeComposeEdit("normalize");

export default function epic(type, work, config) {
  const {
    request = emptyObservableFn,
    success = emptyObservableFn,
    fail = emptyObservableFn,
    wrapper = identity,
    filter = identity,
    normalize = identity
  } = asyncArc()(isFunction(config) ? config({}) : config);

  return (action$, store, deps) => {
    const baseStream = filter(action$.ofType(type), { action$, store, deps });

    return baseStream.flatMap(action => {
      const id = R.path(["meta", "async", "id"], action) || deps.async.uuid();
      const config = { action, action$, id, deps, store };

      const mapMeta = type => action =>
        R.assocPath(["meta", "async"], { id, type }, action);

      return Observable.merge(
        request(config).map(mapMeta("request")),
        wrapper(
          work(config)
            .map(data => normalize(data, config))
            .flatMap(data => success(data, config).map(mapMeta("success"))),
          config
        ).catch(error => fail(error, config).map(mapMeta("fail")))
      );
    });
  };
}
