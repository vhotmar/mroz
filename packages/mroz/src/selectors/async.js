export const state = (state, id) => state.async.byId[id];
