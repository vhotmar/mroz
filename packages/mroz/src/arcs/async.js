import actions from "./actions";
import { async } from "../actions";

export default function asyncArc() {
  return actions(
    ({ id }) => async.request(id),
    (data, { id }) => async.success(id, data),
    (error, { id }) => async.fail(id, error)
  );
}
