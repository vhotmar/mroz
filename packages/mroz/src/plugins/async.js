import R from "ramda";
import { combineReducers } from "redux";
import { handleActions } from "redux-actions";

import { addReducers } from "./utils";

import { async as actions } from "../actions";

export const reducer = handleActions(
  {
    [actions.request]: (state, action) => {
      const id = action.payload.id;
      const prev = state[id] || {};

      return {
        ...state,
        [id]: {
          isFinished: prev.isFinished ? prev.isFinished : false,
          isPending: true,
          error: prev ? prev.error : undefined,
          data: prev.data
        }
      };
    },
    [actions.success]: (state, action) => {
      const id = action.payload.id;

      return {
        ...state,
        [id]: {
          isFinished: true,
          isPending: false,
          error: undefined,
          data: action.payload.data
        }
      };
    },
    [actions.fail]: (state, action) => {
      const id = action.payload.id;
      const prev = state[id];

      return {
        ...state,
        [id]: {
          ...prev,
          isFinished: true,
          isPending: false,
          error: action.payload.error
        }
      };
    }
  },
  {}
);

export default () =>
  R.compose(
    addReducers({
      async: combineReducers({
        byId: reducer
      })
    })
  );
