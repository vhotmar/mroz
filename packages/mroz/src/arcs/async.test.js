import async from "./async";
import { Observable } from "rxjs";
import asyncTest from "../epics/async.test.utils";

describe("async arc", () => {
  it("should dispatch actions", () => {
    const config = async();

    asyncTest({
      config,
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    });

    asyncTest({
      work: () => Observable.throw("error"),
      config,
      callback: res => {
        expect(res).toMatchSnapshot();
      }
    });
  });
});
