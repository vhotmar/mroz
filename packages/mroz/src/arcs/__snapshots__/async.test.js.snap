// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async arc should dispatch actions 1`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "payload": Object {
      "data": Object {
        "data": "foo",
      },
      "id": "id",
    },
    "type": "@@mroz/ASYNC_SUCCESS",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "success",
      },
    },
    "payload": Object {
      "data": Object {
        "data": "foo",
      },
      "id": "id",
    },
    "type": "@@mroz/ASYNC_SUCCESS",
  },
]
`;

exports[`async arc should dispatch actions 2`] = `
Array [
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "request",
      },
    },
    "payload": Object {
      "id": "id",
    },
    "type": "@@mroz/ASYNC_REQUEST",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "fail",
      },
    },
    "payload": Object {
      "error": "error",
      "id": "id",
    },
    "type": "@@mroz/ASYNC_FAIL",
  },
  Object {
    "meta": Object {
      "async": Object {
        "id": "id",
        "type": "fail",
      },
    },
    "payload": Object {
      "error": "error",
      "id": "id",
    },
    "type": "@@mroz/ASYNC_FAIL",
  },
]
`;
